Fix/issue 10113 embeddings use non default tokenizer

camfarineau opened the pull request:
## Title

Embeddings: Not use default `gpt-3.5-turbo`'s tokenizer

## Relevant issues

Fixes https://github.com/BerriAI/litellm/issues/10113

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🐛 Bug Fix

## Changes
- embeddings: allow for passthrough of list of lists of tokens to self hosted vllm models

![litellm_issue_10113_unit_test_local](https://github.com/user-attachments/assets/5ed1e3d7-794b-472f-8204-dc655b740843)

camillefarineau-dg committed:
fix(embeddings): use non default tokenizer when passing list of lists of tokens (int)

camillefarineau-dg committed:
feat(embeddings): allow for passthrough of list of lists of tokens to hosted_vllm models

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=10629) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=10629) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=10629) it.</sub>

krrishdholakia commented on litellm/proxy/proxy_server.py:
```diff
@@ -3814,14 +3814,15 @@ async def embeddings(  # noqa: PLR0915
                         for i in data["input"]:
                             input_list.append(
-                                litellm.decode(model="gpt-3.5-turbo", tokens=i)
+                                litellm.decode(model=m["model_name"], tokens=i)
```
this is dangerous - as a model_name can be anything a user wants to publicly call it - e.g. `gpt-my-new-model` is a valid model_name. 

For your issue - it seems like just handling vllm supports an array of tokens is sufficient - so let's scope this PR to that.

krrishdholakia commented on litellm/proxy/proxy_server.py:
```diff
@@ -3814,14 +3814,15 @@ async def embeddings(  # noqa: PLR0915
                     if m["model_name"] == data["model"] and (
                         m["litellm_params"]["model"] in litellm.open_ai_embedding_models
                         or m["litellm_params"]["model"].startswith("azure/")
+                        or m["litellm_params"]["model"].startswith("hosted_vllm/")
```
add a unit test for this in `test_proxy_server.py` in `tests/litellm` - https://github.com/BerriAI/litellm/blob/e2a9cd0c75f357feef45f90b15d5994cee28ea63/tests/litellm/proxy/test_proxy_server.py#L4

So there are no future regressions.

camillefarineau-dg committed:
Revert "fix(embeddings): use non default tokenizer when passing list of lists of tokens (int)"

This reverts commit a48acd95f860c4fc85853e20668eabffff07cae7.

camillefarineau-dg committed:
refactor(embeddings): use a list to verify if provider accept as input a list of tokens

camillefarineau-dg committed:
fix(embeddings): verify the model name before validating if provider accept a arrays of tokens as input

When passing a list of tokens as input, verify the provider of the model by going through the list of models (`llm_model_list`). First, it check for model name then get the provider and verify if it accept or not arrays of tokens. If yes, then pass, else decode.
Previously, it was verifying provider and model name at the same time resulting in decoding even if the current model checked was not the target one (looping onto `llm_model_list`)

krrishdholakia commented on litellm/proxy/proxy_server.py:
```diff
@@ -3808,24 +3809,26 @@ async def embeddings(  # noqa: PLR0915
-                            )
-                        data["input"] = input_list
-                        break
+                    if m["model_name"] == data["model"]:
```
great! Now let's add a simple test in `test_proxy_server.py` for this, so we don't have regressions and we should be good to merge! 

https://github.com/BerriAI/litellm/blob/0eb0cf4515d505fd840c246ac7e62595b0c39f39/tests/litellm/proxy/test_proxy_server.py#L4

camfarineau replied:
What do you want as a test?
Shouldn't be something held with the `proxy_unit_tests`? https://github.com/BerriAI/litellm/blob/e865a4cddae5c05115834c598030f85e4d90e01d/tests/proxy_unit_tests/test_proxy_server.py#L528

Or should I move this particular code to a proper function in order to write a unit test just for this?

krrishdholakia replied:
we're moving all mock testing inside `tests/litellm` as the other folders contain a mix of mock + real api calls - which can be hard for contributors to run

krrishdholakia replied:
Our github actions for contributor pr's also run `tests/litellm` 
<img width="1328" alt="Screenshot 2025-05-12 at 4 05 31 PM" src="https://github.com/user-attachments/assets/fa1c00e5-a4db-4144-82b2-54561c508599" />

camfarineau replied:
Ok. Just added a test. Let me know if we are good :)

camillefarineau-dg committed:
test(embedding): add unit test to bypass decode for some providers with input as array of tokens

Ref: https://github.com/BerriAI/litellm/issues/10113

camfarineau commented:
Screenshot of the new test passing locally
![litellm_issue_10113_unit_test_local](https://github.com/user-attachments/assets/6de5a1cd-86ac-4536-a494-664a198bcb0f)
